### PR TITLE
Very WIP: Preview: CMake integration

### PR DESCRIPTION
**Pull request by krf**
Heya. For our company project (https://github.com/KDAB/KDStateMachineEditor) we're using libcgraph to layout graphs (via the dot plugin).

I needed make the Windows setup of Graphviz as simple as possible, thus decided to go for CMake files in order to instrument the build for the Visual Studio compiler. I've seen older posts on the Graphviz ML where people were requesting a more straight-forward build system for Graphviz across all platforms. The build instructions for Windows seem overly complicated + outdated, or are even non-functional at that point.

I've hacked together a CMake build system, this is the very-WIP state. The reason I'm putting this on review is simple: I'd like to know if there is _any_ interested on the Graphviz side to move to the CMake build system in the midterm, or if this is completely out of scope.

_If_ you consider replacing your autotools-based system with CMake, then I'm willing to put more effort into extending the current implementation (adding more feature checks, adding more platform checks, etc., etc.). If not, then I don't want to waste time.

Current state:
- Builds all libs _statically_ (bad, I know, but I run into cyclic dependency issues when doing otherwise, didn't completely grasp the Graphviz structure yet)
- Builds fine on Windows + Linux (ensured by internal CI), OS X is failing due to unresolved external symbols (minor issue, missing link libraries, fixable)
- Can link against the dot plugin statically within KDStateMachineEditor (cgraph usage works fine)

I have lots of ideas where I could improve your current code base (import/export handling, feature toggling, etc.), lots of them could be fixed by leveraging features builtin to CMake.

tl;dr: Again, if there's interest I'm willing to spend some time on porting towards CMake (can't promise any ETA, though)

### CODE REVIEW

**Comment by ellson** [comment]
Is there any chance that the CMake support could be added to the code base in parallel with the autotools code?   And would you be willing to maintain it?

For Linux,  static-only is a non-starter, we must have dynamic loading,  but we also need the options for static builds, and in-tree execution for testing purposes.  And the plugin mechanism depends on dynamic loading,  otherwise the executable becomes massive.  If these options could all be supported by CMake than perhaps we could consider moving to it as a primary build system.

**Comment by ellson** [comment]
How is this going?   I doesn't quite look ready for merging yet?
   - the patch commits a generated file:  config.h.in
   - the patch seems to contain changes unrelated to cmake, e.g. lib/cdt/cdh.h, cmd/gvmap,gvmap.c ...
       is this because your branch is missing recent updates from the main branch?

**Comment by krf** [comment]
Not ready for merging, indeed. I don't have a whole lot of time to work on this unfortunately.

Some of the changes in C files were definitely needed in order to get Graphviz built, though I tried to keep the changes in those files as minimal as possible.

I'll try to incorporate your recent changes in the master branch ASAP.

**Comment by ellson** [comment]
pull request declined in current form.

**Comment by krf** [comment]
With #1201 Graphviz now has a CMake-based build system. Great!

**Comment by ErwinJanssen** [comment]
It is still a work in progress so not all libraries and executables are added at the moment, but a full CMake build is coming.
